sql/src/test/java/io/crate/operation/projectors/FetchProjectorTest.java

```diff
@@ -0,0 +1,190 @@
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.projectors;
+
+import com.carrotsearch.hppc.IntHashSet;
+import com.carrotsearch.hppc.IntObjectHashMap;
+import com.carrotsearch.hppc.IntObjectMap;
+import com.carrotsearch.hppc.IntSet;
+import com.google.common.collect.Iterables;
+import com.google.common.util.concurrent.SettableFuture;
+import io.crate.analyze.symbol.FetchReference;
+import io.crate.analyze.symbol.InputColumn;
+import io.crate.analyze.symbol.Reference;
+import io.crate.analyze.symbol.Symbol;
+import io.crate.core.collections.Bucket;
+import io.crate.core.collections.CollectionBucket;
+import io.crate.core.collections.Row;
+import io.crate.core.collections.Row1;
+import io.crate.metadata.*;
+import io.crate.metadata.table.ColumnPolicy;
+import io.crate.operation.RowUpstream;
+import io.crate.operation.projectors.fetch.FetchProjector;
+import io.crate.operation.projectors.fetch.FetchProjectorContext;
+import io.crate.operation.projectors.fetch.TransportFetchOperation;
+import io.crate.planner.node.fetch.FetchSource;
+import io.crate.test.integration.CrateUnitTest;
+import io.crate.testing.CollectingRowReceiver;
+import io.crate.testing.TestingHelpers;
+import io.crate.types.LongType;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+import static org.hamcrest.core.Is.is;
+import static org.mockito.Mockito.*;
+
+public class FetchProjectorTest extends CrateUnitTest {
+
+    private static final TableIdent USER_TABLE_IDENT = new TableIdent(Schemas.DEFAULT_SCHEMA_NAME, "users");
+    private ThreadPool threadPool;
+
+    @Before
+    public void before() throws Exception {
+        threadPool = new ThreadPool("testThreadPool");
+    }
+
+    @After
+    public void after() throws Exception {
+        threadPool.shutdown();
+        threadPool.awaitTermination(1, TimeUnit.SECONDS);
+    }
+
+    @Test
+    public void testFetchSize() throws Throwable {
+        CollectingRowReceiver rowReceiver = new CollectingRowReceiver();
+
+        TransportFetchOperation transportFetchOperation = mock(TransportFetchOperation.class);
+        SettableFuture<IntObjectMap<? extends Bucket>> future = SettableFuture.create();
+        when(transportFetchOperation.fetch(anyString(), any(IntObjectMap.class), anyBoolean()))
+            .thenReturn(future);
+        RowUpstream upstream = mock(RowUpstream.class);
+
+        int fetchSize = 3;
+        FetchProjector fetchProjector = prepareFetchProjector(fetchSize, rowReceiver, transportFetchOperation, upstream);
+
+        IntObjectMap<Bucket> intObjectMap = new IntObjectHashMap<>(2);
+        Collection<Object[]> rows = new ArrayList<>(10);
+
+        int i;
+        for (i = 1; i <= 10; i++) {
+            Row row = new Row1((long)i);
+            rows.add(new Object[]{row});
```

Comment:
This seems a bit odd.  Either rows should contain just Row instances or Object arrays that contain the values, but not other rows.

You could use:
```
Row row = new Row1(i);
rows.add(row.materialize());
```